Objects/complexobject.c

```diff
@@ -408,11 +409,9 @@ complex_hash(PyComplexObject *v)
 {
     Py_uhash_t hashreal, hashimag, combined;
     hashreal = (Py_uhash_t)_Py_HashDouble((PyObject *) v, v->cval.real);
-    if (hashreal == (Py_uhash_t)-1)
-        return -1;
+    assert(hashreal != (Py_uhash_t)-1);
```

Comment:
This works because `_Py_HashDouble` cannot fail (and hence return a minus -1). The fact that _Py_HashDouble cannot fail however is an implementation detail that I could not find documented.

The assert is good to add, but I still feel a bit uncomfortable with this change. At least a comment would be in place

Reply:
In fact, it's documented, e.g. here:
https://docs.python.org/3/library/stdtypes.html#hashing-of-numeric-types
(see python code examples)

I don't think comment here does make sense, asserts already tell us everything: ``_Py_HashDouble()`` shouldn't return ``-1``.

Reply:
I was worried that `_Py_HashDouble` could return -1 in case of a memory error (the documentation makes no claims about that, only that -1 is reserved) since the method takes a pointer to a python object. However, the `_Py_HashDouble` only uses the Python object with  `_Py_HashPointer` which cannot fail. The assert is fine.

Reply:
I agree with @eendebakpt: I'd also like to see a comment here, for readability. Without this comment, someone reading the `complexobject.c` code has to go elsewhere to find out why the assert is valid; the comment helps explain the 'why' of the code.

Reply:
Comment was added.